
package iosig;

option java_package = "edu.iit.iosig";
option java_outer_classname = "IOSignatureList";

message Signature {
  enum SigType {
    CONTIGUOUS = 0;
    FIXED_STRIDED = 1;
    TWOD_STRIDED = 2;
    THREED_STRIDED = 3;
    MARKOV = 4;
  }
  required int32 id = 1;
  required SigType type = 2;
  required int32 number_of_access = 3;
  required int32 initial_position = 4;
  required int32 initial_size = 5;

  message Pattern_Basic {
    required int32 initial_position = 1;
    required int32 dimension = 2;
    required int32 increment = 3;
    required int32 repetition = 4;
  }

  message Pattern_Advanced {
    required Pattern_Basic offset_pattern = 1;
    required Pattern_Basic size_pattern = 2;
    required Pattern_Basic repetition_pattern = 3;
  }

  message Pattern_2dBasic {
    required int32 initial_position = 1;
    required int32 dimension = 2;
    required Pattern_Basic increment = 3;
    required int32 repetition = 4;
  }

  message Pattern_2dAdvanced {
    required Pattern_2dBasic offset_pattern = 1;
    required Pattern_2dBasic size_pattern = 2;
    required Pattern_Basic repetition_pattern = 3;
  }
  optional Pattern_Basic basic_pattern = 6;
  optional Pattern_Advanced adv_pattern = 7;
  optional Pattern_2dBasic twodbasic_pattern = 8;
  optional Pattern_2dAdvanced twodadv_pattern = 9;
}

message SignatureList {
  required string trace = 1;
  repeated Signature sig = 2;
}